frontend/public/components/storage-class-form.tsx

```diff
@@ -0,0 +1,927 @@
+    this.setState({error: null, fieldErrors: {}}, function() {
+      let fieldErrors = that.state.fieldErrors;
+      let validationSuccessful = true;
+
+      const nameValidation = that.validateName();
+      if (!nameValidation.nameIsValid) {
+        fieldErrors.nameValidationMsg = nameValidation.error;
+        validationSuccessful = false;
+      }
+
+      if (that.state.newStorageClass.type === null) {
+        validationSuccessful = false;
+      } else if (!that.validateParameters()) {
+        validationSuccessful = false;
+      }
+
+      if (!that.allRequiredFieldsFilled()) {
+        validationSuccessful = false;
+      }
+
+      this.setState({fieldErrors, validationSuccessful});
+    });
+  };
+
+  validateName = () => {
+    const updatedName = this.state.newStorageClass.name;
+    const nameUpdated = updatedName !== this.previousName;
+    const returnVal = {
+      error: null,
+      nameIsValid: true
+    };
+
+    if (nameUpdated) {
+      if (updatedName.trim().length === 0) {
+        returnVal.error = 'Storage name is required';
+        returnVal.nameIsValid = false;
+      } else if (this.resources) {
+        for (let i = 0; i < this.resources.data.length; i++) {
+          if (this.resources.data[i].metadata.name === updatedName.toLowerCase()) {
+            returnVal.error = 'Storage name must be unique';
+            returnVal.nameIsValid = false;
+          }
+        }
+      }
+
+      this.previousName = updatedName;
+    }
+
+    return returnVal;
+  };
+
+  validateParameters = () => {
+    let validationSuccessful = true;
+    const params = this.state.newStorageClass.parameters;
+    for (let key in params) {
+      if (params.hasOwnProperty(key) && params[key].validationMsg !== null) {
+        validationSuccessful = false;
+        break;
+      }
+    }
+
+    return validationSuccessful;
+  };
+
+  allRequiredFieldsFilled = () => {
+    if (this.state.newStorageClass.name.trim().length === 0) {
+      return false;
+    }
+
+    const storageType = this.state.newStorageClass.type;
+    if (storageType === null) {
+      return false;
+    }
+
+    const allParamsForType = this.storageTypes[storageType].parameters;
+    const userEnteredParams = this.state.newStorageClass.parameters;
+    let allReqdFieldsEntered = true;
+
+    for (let key in allParamsForType) {
+      if (allParamsForType.hasOwnProperty(key)) {
+        const value = _.get(userEnteredParams, `${key}.value`, null);
+        const valueExists = value !== null && value !== '' && !_.isEmpty(value);
+
+        if (this.paramIsRequired(key) && !valueExists) {
+          allReqdFieldsEntered = false;
+          break;
+        }
+      }
+    }
+
+    return allReqdFieldsEntered;
+  };
+
+  paramIsRequired = (paramKey, params = this.state.newStorageClass.parameters) => {
+    const requiredParam = _.get(this.storageTypes[this.state.newStorageClass.type], ['parameters', paramKey, 'required'], null);
```

Comment:
Shouldn't need to default to `null` here

Reply:
If I don't default to null, I get a typescript warning about a possibly undefined variable.